### ISSUE TITLE
Model parallelism

### ISSUE THREAD

**Issue by KelleyYin** [pytorch/fairseq#350]
I want to use model parallelism rather than data parallelism for this program.
How to do it ?

**Comment by edunov** [comment]
What's your use case? 
I'm planning to explore model parallelism to see how far we can push these models. And it should be possible to do it with a relatively small change. But there is no built-in support yet.

**Comment by edunov** [comment]
Do you train them separately or jointly? Do they see same mini-batches during training? Is the loss shared? 

Details depend on what exactly you do with these models. But yes, you can put them on different GPU's. You'll have to remove lines like this: 

https://github.com/fairinternal/fairseq-py/blob/master/train.py#L31
where we make an assumption of 1 GPU per worker

And then you will have to manually put your models on the correct GPUs. Currently, we just call model.cuda(), but you can also pass device=X to this function. 

Finally, you'll need to stitch everything together, e.g. make sure input data is copied to where it should be. Make sure the output is copied where it should be. Make sure the backward pass is handled correctly etc.

**Comment by KelleyYin** [comment]
@edunov 
Two model will are jointly trained.  Total loss consists of two parts, one part is shared and other part is independent.
The link your provided above is invalid .Can you provide again ?  
Thanks a lot .

**Comment by KelleyYin** [comment]
@carlosep93 
You can find the solution from [here](https://discuss.pytorch.org/t/how-to-implement-model-parallelism/29285).
